# HR Analytics

About Practice Problem: HR Analytics
HR analytics is revolutionising the way human resources departments operate, leading to higher efficiency and better results overall. Human resources has been using analytics for years. However, the collection, processing and analysis of data has been largely manual, and given the nature of human resources dynamics and HR KPIs, the approach has been constraining HR. Therefore, it is surprising that HR departments woke up to the utility of machine learning so late in the game. Here is an opportunity to try predictive analytics in identifying the employees most likely to get promoted.

# Client

Can you predict which employee has a higher chance of getting promoted and help the help the organization to expedite the appraisal and promotion process?


In [95]:
import pandas as pd
import numpy as np

# Loading Data from CSV File   --->  Step 1

We load two CSV files 

1. Test  Data  ---> Test.CSV 
2. Sample Data ---> Train.CSV

loading Training & test data





In [96]:

test_data=pd.read_csv(r'E:\Phyton Data\test.csv',index_col=0,
                       header=0)
train_data=pd.read_csv(r'E:\Phyton Data\train.csv',index_col=0,header=0)

# Preprocessing the training dataset

1 Test Data


# 1 Test Data

In [97]:
test_data.head()

department     region   education gender  \
employee_id                                                    
8724                Technology  region_26  Bachelor's      m   
74430                       HR   region_4  Bachelor's      f   
72255        Sales & Marketing  region_13  Bachelor's      m   
38562              Procurement   region_2  Bachelor's      f   
64486                  Finance  region_29  Bachelor's      m   

            recruitment_channel  no_of_trainings  age  previous_year_rating  \
employee_id                                                                   
8724                   sourcing                1   24                   NaN   
74430                     other                1   31                   3.0   
72255                     other                1   31                   1.0   
38562                     other                3   31                   2.0   
64486                  sourcing                1   30                   4.0   

             length_of_service  KPIs_met >80%  awards_won?  avg_training_score  
employee_id                                                                     
8724                         1              1            0                  77  
74430                        5              0            0                  51  
72255                        4              0            0                  47  
38562                        9              0            0                  65  
64486                        7              0            0                  61

# 2 Train Data 

In [98]:
train_data.head()

department     region         education gender  \
employee_id                                                          
65438        Sales & Marketing   region_7  Master's & above      f   
65141               Operations  region_22        Bachelor's      m   
7513         Sales & Marketing  region_19        Bachelor's      m   
2542         Sales & Marketing  region_23        Bachelor's      m   
48945               Technology  region_26        Bachelor's      m   

            recruitment_channel  no_of_trainings  age  previous_year_rating  \
employee_id                                                                   
65438                  sourcing                1   35                   5.0   
65141                     other                1   30                   5.0   
7513                   sourcing                1   34                   3.0   
2542                      other                2   39                   1.0   
48945                     other                1   45                   3.0   

             length_of_service  KPIs_met >80%  awards_won?  \
employee_id                                                  
65438                        8              1            0   
65141                        4              0            0   
7513                         7              0            0   
2542                        10              0            0   
48945                        2              0            0   

             avg_training_score  is_promoted  
employee_id                                   
65438                        49            0  
65141                        60            0  
7513                         50            0  
2542                         50            0  
48945                        73            0

# Shape the Data Base 1.Test DataBase 2. Train Database


In [99]:
print(train_data.shape)
print(test_data.shape)

(54808, 13)
(23490, 12)


# Train having Samples---> 54808, Variables --->  13

# Test having Samples---> 23490, Variables ---> 12
 

In [100]:
train_data.head()

department     region         education gender  \
employee_id                                                          
65438        Sales & Marketing   region_7  Master's & above      f   
65141               Operations  region_22        Bachelor's      m   
7513         Sales & Marketing  region_19        Bachelor's      m   
2542         Sales & Marketing  region_23        Bachelor's      m   
48945               Technology  region_26        Bachelor's      m   

            recruitment_channel  no_of_trainings  age  previous_year_rating  \
employee_id                                                                   
65438                  sourcing                1   35                   5.0   
65141                     other                1   30                   5.0   
7513                   sourcing                1   34                   3.0   
2542                      other                2   39                   1.0   
48945                     other                1   45                   3.0   

             length_of_service  KPIs_met >80%  awards_won?  \
employee_id                                                  
65438                        8              1            0   
65141                        4              0            0   
7513                         7              0            0   
2542                        10              0            0   
48945                        2              0            0   

             avg_training_score  is_promoted  
employee_id                                   
65438                        49            0  
65141                        60            0  
7513                         50            0  
2542                         50            0  
48945                        73            0

In [101]:
test_data.head()

department     region   education gender  \
employee_id                                                    
8724                Technology  region_26  Bachelor's      m   
74430                       HR   region_4  Bachelor's      f   
72255        Sales & Marketing  region_13  Bachelor's      m   
38562              Procurement   region_2  Bachelor's      f   
64486                  Finance  region_29  Bachelor's      m   

            recruitment_channel  no_of_trainings  age  previous_year_rating  \
employee_id                                                                   
8724                   sourcing                1   24                   NaN   
74430                     other                1   31                   3.0   
72255                     other                1   31                   1.0   
38562                     other                3   31                   2.0   
64486                  sourcing                1   30                   4.0   

             length_of_service  KPIs_met >80%  awards_won?  avg_training_score  
employee_id                                                                     
8724                         1              1            0                  77  
74430                        5              0            0                  51  
72255                        4              0            0                  47  
38562                        9              0            0                  65  
64486                        7              0            0                  61

In [126]:
test_data.head()

department  region  education  gender  recruitment_channel  \
employee_id                                                               
8724                  8      18          0       1                    2   
74430                 2      28          0       0                    0   
72255                 7       4          0       1                    0   
38562                 5      11          0       0                    0   
64486                 1      21          0       1                    2   

             no_of_trainings  age  previous_year_rating  length_of_service  \
employee_id                                                                  
8724                       1   24                   3.0                  1   
74430                      1   31                   3.0                  5   
72255                      1   31                   1.0                  4   
38562                      3   31                   2.0                  9   
64486                      1   30                   4.0                  7   

             KPIs_met >80%  awards_won?  avg_training_score  
employee_id                                                  
8724                     1            0                  77  
74430                    0            0                  51  
72255                    0            0                  47  
38562                    0            0                  65  
64486                    0            0                  61

In [127]:
train_data.head()

department  region  education  gender  recruitment_channel  \
employee_id                                                               
65438                 7      31          2       0                    2   
65141                 4      14          0       1                    0   
7513                  7      10          0       1                    2   
2542                  7      15          0       1                    0   
48945                 8      18          0       1                    0   

             no_of_trainings  age  previous_year_rating  length_of_service  \
employee_id                                                                  
65438                      1   35                   5.0                  8   
65141                      1   30                   5.0                  4   
7513                       1   34                   3.0                  7   
2542                       2   39                   1.0                 10   
48945                      1   45                   3.0                  2   

             KPIs_met >80%  awards_won?  avg_training_score  is_promoted  
employee_id                                                               
65438                    1            0                  49            0  
65141                    0            0                  60            0  
7513                     0            0                  50            0  
2542                     0            0                  50            0  
48945                    0            0                  73            0

# Finding the missing values  --> Test Database 


In [104]:
test_data.isnull().sum()

department                 0
region                     0
education               1034
gender                     0
recruitment_channel        0
no_of_trainings            0
age                        0
previous_year_rating    1812
length_of_service          0
KPIs_met >80%              0
awards_won?                0
avg_training_score         0
dtype: int64

# Finding the missing values --> Train Database 

In [105]:
train_data.isnull().sum()

department                 0
region                     0
education               2409
gender                     0
recruitment_channel        0
no_of_trainings            0
age                        0
previous_year_rating    4124
length_of_service          0
KPIs_met >80%              0
awards_won?                0
avg_training_score         0
is_promoted                0
dtype: int64

# Missing Values Replace with Zero

# importing categorical missing data with mode value

# importing Numrical  missing data with mean value

In [106]:
colname1=["education"]

for x in colname1:
    test_data[x].fillna(test_data[x].mode()[0],inplace=True)
    

In [107]:
colname1=["previous_year_rating"]

for x in colname1:
    test_data[x].fillna(round(train_data[x].mean(),0),inplace=True)

In [108]:
test_data.isnull().sum()

department              0
region                  0
education               0
gender                  0
recruitment_channel     0
no_of_trainings         0
age                     0
previous_year_rating    0
length_of_service       0
KPIs_met >80%           0
awards_won?             0
avg_training_score      0
dtype: int64

In [109]:
colname1=["education"]

for x in colname1:
    train_data[x].fillna(train_data[x].mode()[0],inplace=True)

In [110]:
colname1=["previous_year_rating"]

for x in colname1:
    train_data[x].fillna(round(train_data[x].mean(),0),inplace=True)

In [111]:
train_data.isnull().sum()

department              0
region                  0
education               0
gender                  0
recruitment_channel     0
no_of_trainings         0
age                     0
previous_year_rating    0
length_of_service       0
KPIs_met >80%           0
awards_won?             0
avg_training_score      0
is_promoted             0
dtype: int64

# Convert Alphanumric to Numaric


In [112]:
Colname=[]
for x in train_data.columns[:]:
    if train_data[x].dtype=='object':
        Colname.append(x)
Colname

['department', 'region', 'education', 'gender', 'recruitment_channel']

In [113]:
#transforming categorical data to numerical
from sklearn import preprocessing
colname=['department', 'region','education','gender','recruitment_channel']

le=preprocessing.LabelEncoder()

for x in colname:
    train_data[x]=le.fit_transform(train_data[x])

In [114]:
train_data.head()

department  region  education  gender  recruitment_channel  \
employee_id                                                               
65438                 7      31          2       0                    2   
65141                 4      14          0       1                    0   
7513                  7      10          0       1                    2   
2542                  7      15          0       1                    0   
48945                 8      18          0       1                    0   

             no_of_trainings  age  previous_year_rating  length_of_service  \
employee_id                                                                  
65438                      1   35                   5.0                  8   
65141                      1   30                   5.0                  4   
7513                       1   34                   3.0                  7   
2542                       2   39                   1.0                 10   
48945                      1   45                   3.0                  2   

             KPIs_met >80%  awards_won?  avg_training_score  is_promoted  
employee_id                                                               
65438                    1            0                  49            0  
65141                    0            0                  60            0  
7513                     0            0                  50            0  
2542                     0            0                  50            0  
48945                    0            0                  73            0

In [115]:
Colname=[]
for x in test_data.columns[:]:
    if test_data[x].dtype=='object':
        Colname.append(x)
Colname

['department', 'region', 'education', 'gender', 'recruitment_channel']

In [116]:
#transforming categorical data to numerical
from sklearn import preprocessing
colname=['department', 'region','education','gender','recruitment_channel']

le=preprocessing.LabelEncoder()

for x in colname:
    test_data[x]=le.fit_transform(test_data[x])

In [117]:
test_data.head()

department  region  education  gender  recruitment_channel  \
employee_id                                                               
8724                  8      18          0       1                    2   
74430                 2      28          0       0                    0   
72255                 7       4          0       1                    0   
38562                 5      11          0       0                    0   
64486                 1      21          0       1                    2   

             no_of_trainings  age  previous_year_rating  length_of_service  \
employee_id                                                                  
8724                       1   24                   3.0                  1   
74430                      1   31                   3.0                  5   
72255                      1   31                   1.0                  4   
38562                      3   31                   2.0                  9   
64486                      1   30                   4.0                  7   

             KPIs_met >80%  awards_won?  avg_training_score  
employee_id                                                  
8724                     1            0                  77  
74430                    0            0                  51  
72255                    0            0                  47  
38562                    0            0                  65  
64486                    0            0                  61

# Definding Variable:
Train Data & Test Data
X is Independent Variable Y is Dependent Variable

In [134]:
X_train=train_data.values[:,:-1]
Y_train=train_data.values[:,-1]
Y_train=Y_train.astype(int)

In [136]:

X_test=test_data.values[:,:]

In [139]:
print(X_test)
#print(Y_train)

[[ 8. 18.  0. ...  1.  0. 77.]
 [ 2. 28.  0. ...  0.  0. 51.]
 [ 7.  4.  0. ...  0.  0. 47.]
 ...
 [ 2.  7.  0. ...  0.  0. 50.]
 [ 5. 24.  0. ...  0.  0. 70.]
 [ 8.  8.  2. ...  1.  0. 89.]]


# Scaler

In [140]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)
print(X_train)


[[ 0.80893285  1.77652645  1.61975831 ...  1.35687789 -0.15401776
  -1.07593145]
 [-0.38818322 -0.16303375 -0.62883817 ... -0.73698599 -0.15401776
  -0.25328242]
 [ 0.80893285 -0.61940086 -0.62883817 ... -0.73698599 -0.15401776
  -1.00114517]
 ...
 [-1.98433798 -1.76031862 -0.62883817 ...  1.35687789 -0.15401776
   1.1676568 ]
 [ 0.80893285  2.00471    -0.62883817 ... -0.73698599 -0.15401776
  -1.37507655]
 [-1.1862606  -0.16303375 -0.62883817 ... -0.73698599 -0.15401776
  -1.07593145]]


# Accuracy Test SVM 1

In [142]:
from sklearn import svm
classifier=svm.SVC(kernel='rbf',C=1.0,gamma=0.1)

#performing kfold_cross_validation
from sklearn.model_selection import KFold
kfold_cv=KFold(n_splits=10)
print(kfold_cv)

from sklearn.model_selection import cross_val_score
#running the model using scoring metric as accuracy
kfold_cv_result=cross_val_score(estimator=classifier,X=X_train,
y=Y_train, cv=kfold_cv)
print(kfold_cv_result)
#finding the mean
print(kfold_cv_result.mean())

KFold(n_splits=10, random_state=None, shuffle=False)
[0.91862799 0.92081737 0.92811531 0.92921    0.92756796 0.92154716
 0.9231892  0.92081737 0.92664234 0.91879562]
0.9235330311613976


# Accuracy Test SVM 2

In [143]:
classifier=svm.SVC(kernel='rbf',C=10.0,gamma=0.001)

#performing kfold_cross_validation
from sklearn.model_selection import KFold
kfold_cv=KFold(n_splits=10)
print(kfold_cv)

from sklearn.model_selection import cross_val_score
#running the model using scoring metric as accuracy
kfold_cv_result=cross_val_score(estimator=classifier,X=X_train,
y=Y_train, cv=kfold_cv)
print(kfold_cv_result)
#finding the mean
print(kfold_cv_result.mean())

KFold(n_splits=10, random_state=None, shuffle=False)
[0.91552636 0.91552636 0.92483124 0.9235541  0.92410144 0.91625616
 0.91954023 0.91534392 0.92153285 0.91587591]
0.9192088562079753


# Running LogisticRegression to See the accuracy

In [144]:
from sklearn.linear_model import LogisticRegression

classifier=(LogisticRegression())

from sklearn.model_selection import KFold
kfold_cv=KFold(n_splits=10)
print(kfold_cv)

from sklearn.model_selection import cross_val_score
#running the model using scoring metric as accuracy
kfold_cv_result=cross_val_score(estimator=classifier,X=X_train,
y=Y_train, cv=kfold_cv)
print(kfold_cv_result)
#finding the mean
print(kfold_cv_result.mean())

KFold(n_splits=10, random_state=None, shuffle=False)


C:\Users\Dell\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Dell\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Dell\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Dell\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Dell\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this w

[0.91351943 0.91406678 0.92373654 0.92099982 0.92154716 0.91388433
 0.91771575 0.91315453 0.9189781  0.91386861]
0.9171471053952805
